scenarios/two_party_countersigning/README.md

```diff
@@ -0,0 +1,36 @@
+## remote_call
+
+### Description
+
+This scenario tests the throughput of `remote_call` operations.
+
+**warning** This is a TryCP-based scenario and needs to be run differently to other scenarios.
+
+### Waiting for peer discovery
+
+This scenario reads the environment variable `MIN_PEERS` and waits for at least that many peers to be available before
+starting the agent behaviour. It will wait up to two minutes then proceed regardless.
+
+The scenario is not able to check that you have configured more peers than the minimum you have set, so you should
+ensure that you have configured enough peers to meet the minimum.
+
+Note that the number of peers seen by each node includes itself. So having two nodes means that each node will
+immediately see one peer after app installation.
+
+This configuration defaults to 2 peers.
+
+### Suggested command
+
+You can run the scenario locally with the following command:
+
+```bash
+RUST_LOG=OFF MIN_PEERS=5 cargo run --package two_party_countersigning -- --targets targets-ci.yaml --behaviour initiate:2 --behaviour participate:3 --instances-per-target 5 --duration 300 --reporter influx-file
```

Comment:
I wonder, in light of your comment about conflicts, since (to my understanding) that's not the path we're trying to get performance data about in the test, if we'd rather do some string processing and just have some dedicated groups eliminating the possibility of conflict:

```
--behaviour init-a:1 --behaviour part-a: 3 --behaviour init-b:1 --behaviour part-b: 3, etc
```

Reply:
We could achieve partitioning within the test if we want to, if we expose the number of peers assigned to a behaviour then we can use `%`-logic to partition the targets by their index. I.e. rather than just advertising "I am a participant", they could advertise "I am participant X" and the initiators could select a subset from that by their own index.

I'm not keen to do this though. It effectively becomes equivalent to "--behaviour init:1 --behaviour part:5 --targets 6` with some  noise from syncing ops. At that point the network, or subgraph of the network, is limited by how fast one agent can execute countersigning sessions. At that point, because the accept is much faster than the initiate we may as well just test with two nodes if we want to measure purely throughput. I think that is a valid way to run this test that we could run a report on.

The more general case though is to allow conflicts to happen. On a real network you wouldn't necessarily know that somebody you want to do countersigning with is already doing a session with somebody else. All I'm really worried about with the conflicts is a system issue. If we see the same behaviour on the HPs that I'm seeing locally then something is wrong with the test

Reply:
I had a though overnight about what that might be. The accepting agents must be listening again before their chain is unlocked. This might be fixable by waiting for their chain to be unlocked before resuming listening to signals. It shouldn't backlog signals more than the number of initiating agents because they will always wait for a response. So doing that should lower the throughput and get rid of any errors... I think

Reply:
Yes, that was part of the problem. Digging into Holochain code I found that there is a signal emitted when the session completes. So I've solved one problem with that. Now I have another problem which is that another kind of error was being hidden by the first

Reply:
Awesome, yeah I certainly don't mind having the conflicting sessions in there as long as we can distinguish them in the metrics :+1:

Reply:
Fair point, the error metrics are not very informative at the moment because there are so many possible exit points and I'm really not clear which ones we should care about. Plus the error messages that come back contain far too many strings to be easily processable. I think this will need work when we come to reporting on this. For now all you can really tell is how many sessions were attempted, how many succeeded and how long the successful sessions took. I think that's a useful starting point for now, definitely not reaching what I'd think of as ideal.